src/tokenizer/chain.rs

```diff
@@ -0,0 +1,185 @@
+use tokenizer_api::{Token, TokenStream, Tokenizer};
+
+/// Trait to allow easily chaining two [`Tokenizer`].
+pub trait ChainTokenizerExt: Sized {
+    /// Produce a [`Tokenizer`] which runs through the first tokenizer, and then through the second.
+    fn chain<T: Tokenizer>(self, next: T) -> ChainTokenizer<Self, T>;
+}
+
+impl<F: Tokenizer> ChainTokenizerExt for F {
+    fn chain<T: Tokenizer>(self, second: T) -> ChainTokenizer<Self, T> {
+        ChainTokenizer {
+            first: self,
+            second,
+        }
+    }
+}
+
+/// A tokenizer running through the first tokenizer and then through the second.
+#[derive(Clone)]
+pub struct ChainTokenizer<F, S> {
+    first: F,
+    second: S,
+}
+
+impl<F, S> ChainTokenizer<F, S>
+where
+    F: Tokenizer,
+    S: Tokenizer,
+{
+    /// Create a new tokenzier, chaining the two provided ones.
+    pub fn new(first: F, second: S) -> Self {
+        Self { first, second }
+    }
+}
+
+impl<F, S> Tokenizer for ChainTokenizer<F, S>
+where
+    F: Tokenizer,
+    S: Tokenizer,
+{
+    type TokenStream<'a> = ChainTokenStream<'a, F, S>;
+
+    fn token_stream<'a>(&'a mut self, text: &'a str) -> Self::TokenStream<'a> {
+        ChainTokenStream {
+            state: State::First,
+            first: self.first.token_stream(text),
+            second: self.second.token_stream(text),
+            dummy: Token::default(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum State {
+    First,
+    Second,
+    Done,
+}
+
+pub struct ChainTokenStream<'a, F, S>
+where
+    F: Tokenizer,
+    S: Tokenizer,
+{
+    first: F::TokenStream<'a>,
```

Comment:
I wonder if this could be simplified to

```rust
pub struct ChainTokenStream<F, S> {
  first: Option<F>,
  second: S,
}

impl<F,S> TokenStream for ChainTokenStream<F,S> where F: TokenStream, S: TokenStream {
  fn advance(&mut self) -> bool {
    if let Some(first) =  &mut self.first {
       if first.advance() {
         true
       } else {
         self.first = None; 
         false
       }
    } else {
       self.second.advance()
    }
  }
}
```

or does this run afoul of the hidden lifetime `'a`?